lib/libwireguard.py

```diff
@@ -0,0 +1,140 @@
+            # if not, generate one
+            if self.kv.get("private-key"):
+                log("Private key already configured", level="debug")
+            else:
+                log("Generating private key", level="debug")
+                private_key = self.run_wg(["genkey"])
+                self.kv.set("private-key", private_key)
+
+        if not self.kv.get("public-key"):
+            # Generate a public key for the configured/generated private key
+            public_key = self.run_wg(["pubkey"], private_key.encode())
+            self.kv.set("public-key", public_key)
+
+    def configure(self):
+        """Write configuration for WireGuard."""
+        self.configure_keys()
+        self.configure_forwarding()
+
+        host.service('stop', self.service_name)
+        peers_string = base64.b64decode(self.charm_config["peers"])
+        peers_yaml = yaml.load(peers_string)
+        context = {
+            "private_key": self.kv.get("private_key"),
+            "listen_port": self.charm_config["listen-port"],
+            "peers": peers_yaml,
+            "address": self.charm_config["address"],
+            "forward": self.charm_config["forward-ip"],
+            "forward_dev": self.charm_config["forward-dev"],
+        }
+        templating.render("wg0.conf.j2", self.cfg_file, context, perms=0o660)
+        host.service("enable", self.service_name)
+        host.service("start", self.service_name)
+
+        self.configure_ports()
+        hookenv.open_port(self.charm_config["listen-port"], protocol="UDP")
+
+    def configure_ports(self):
+        """Configure listening ports."""
+        listen_port = self.charm_config["listen-port"]
+        for open_port in hookenv.opened_ports():
+            port, protocol = open_port.split("/")
+            if protocol != 'udp' and port != listen_port:
+                hookenv.close_port(port, protocol=protocol.upper())
+        hookenv.open_port(self.charm_config["listen-port"], protocol="UDP")
+
+    def configure_forwarding(self):
+        """Disable ip forwarding in the kernel."""
+        if self.charm_config['forward-ip']:
+            sysctl_contents = "net.ipv4.ip_forward=1\nnet.ipv6.conf.all.forwarding=1"
+            sysctl_file = open(self.sysctl_file, "w")
+            sysctl_file.write(sysctl_contents)
+            sysctl_file.close()
+        else:
+            os.remove(self.sysctl_file)
+
+        cmd = ["sysctl", "--system"]
+        try:
+            check_call(cmd)
+        except CalledProcessError as e:
+            log("Failed to restore sysctl:", level="error")
+            log("Cmd: {}".format(cmd), level="error")
+            log("Error: {}".format(e.output), level="error")
+
+    def get_config_action(self):
+        """Retrieve and return settings and key data for get-config action."""
+        with open(self.public_key_file, "r") as key:
+            public_key = key.read().strip("\n")
```

Comment:
Doesn't this need to be updated to use the kv store now? I suspect this is going to error on a new install or an upgrade after the migration.